src/Core/State/ProviderInterface.php

```diff
@@ -0,0 +1,35 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\State;
+
+/**
+ * Retrieves data from a persistence layer.
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ * @experimental
+ */
+interface ProviderInterface
+{
+    /**
+     * Provides data.
+     *
+     * @return object|array|null
+     */
+    public function provide(string $resourceClass, array $identifiers = [], array $context = []);
```

Comment:
should we split this into `provide` and `provideAll`? this method can never be statically typed unless a special plugin is made for it.

this would also allow us to add type hint to `provide` ( `?object` ) and to `provideAll` ( `array` ).

Reply:
As we want to remove concepts of collection and item we don't really want to have 2 methods. This should just be typed `: mixed`.

Reply:
Implementations however will be able to provide the appropriate type. We could use generics here to solve the problem.